Content.Shared/Temperature/Systems/SharedTemperatureSystem.cs

```diff
@@ -1,80 +1,431 @@
+        SubscribeLocalEvent<TemperatureProtectionComponent, InventoryRelayedEvent<ModifyChangedTemperatureEvent>>(OnTemperatureChangeAttempt);
+
+        SubscribeLocalEvent<InternalTemperatureComponent, MapInitEvent>(OnInit);
+
+        SubscribeLocalEvent<ChangeTemperatureOnCollideComponent, ProjectileHitEvent>(ChangeTemperatureOnCollide);
+
+        // Allows overriding thresholds based on the parent's thresholds.
+        SubscribeLocalEvent<TemperatureComponent, EntParentChangedMessage>(OnParentChange);
+        SubscribeLocalEvent<ContainerTemperatureDamageThresholdsComponent, ComponentStartup>(OnParentThresholdStartup);
+        SubscribeLocalEvent<ContainerTemperatureDamageThresholdsComponent, ComponentShutdown>(OnParentThresholdShutdown);
+
+        InitializeMoveSpeed();
     }
 
-    private void OnTemperatureChanged(Entity<TemperatureSpeedComponent> ent, ref OnTemperatureChangeEvent args)
+    public override void Update(float frameTime)
     {
-        foreach (var (threshold, modifier) in ent.Comp.Thresholds)
+        base.Update(frameTime);
+
+        // conduct heat from the surface to the inside of entities with internal temperatures
+        var query = EntityQueryEnumerator<InternalTemperatureComponent, TemperatureComponent>();
+        while (query.MoveNext(out var uid, out var comp, out var temp))
         {
-            if (args.CurrentTemperature < threshold && args.LastTemperature > threshold ||
-                args.CurrentTemperature > threshold && args.LastTemperature < threshold)
-            {
-                ent.Comp.NextSlowdownUpdate = _timing.CurTime + SlowdownApplicationDelay;
-                ent.Comp.CurrentSpeedModifier = modifier;
-                Dirty(ent);
+            // don't do anything if they equalised
+            var diff = Math.Abs(temp.CurrentTemperature - comp.Temperature);
+            if (diff < 0.1f)
+                continue;
+
+            // heat flow in W/m^2 as per fourier's law in 1D.
+            var q = comp.Conductivity * diff / comp.Thickness;
+
+            // convert to J then K
+            var joules = q * comp.Area * frameTime;
+            var degrees = joules / GetHeatCapacity((uid, temp, null));
+            if (temp.CurrentTemperature < comp.Temperature)
+                degrees *= -1;
+
+            // exchange heat between inside and surface
+            comp.Temperature += degrees;
+            Dirty(uid,comp);
+            ForceChangeTemperature((uid, temp), temp.CurrentTemperature - degrees);
+        }
+
+        UpdateDamage(frameTime);
+        UpdateMoveSpeed(frameTime);
+    }
+
+    private void UpdateDamage(float frameTime)
+    {
+        _accumulatedFrametime += frameTime;
+
+        if (_accumulatedFrametime < UpdateInterval)
+            return;
+
+        _accumulatedFrametime -= UpdateInterval;
+
+        if (ShouldUpdateDamage.Count == 0)
+            return;
+
+        foreach (var comp in ShouldUpdateDamage)
```

Comment:
```suggestion
        foreach (var ent in ShouldUpdateDamage)
```
rename to avoid confusion